docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md

```diff
@@ -0,0 +1,378 @@
+# FileSystemStore in Mbed OS
+
+- [FileSystemStore in Mbed OS](#filesystemstore-in-mbed-os)
+    + [Revision history](#revision-history)
+- [Introduction](#introduction)
+    + [Overview and background](#overview-and-background)
+    + [Requirements and assumptions](#requirements-and-assumptions)
+- [System architecture and high-level design](#system-architecture-and-high-level-design)
+  * [Design basics](#design-basics)
+- [Detailed design](#detailed-design)
+    + [Class header](#class-header)
+    + [Important data structures](#important-data-structures)
+    + [Initialization and reset](#initialization-and-reset)
+    + [Core APIs](#core-apis)
+    + [Incremental set APIs](#incremental-set-apis)
+    + [Key iterator APIs](#key-iterator-apis)
+- [Usage scenarios and examples](#usage-scenarios-and-examples)
+    + [Standard usage of the class](#standard-usage-of-the-class)
+- [Other information](#other-information)
+    + [Open issues](#open-issues)
+
+
+### Revision history
+
+| Revision 	| Date           	| Authors                                                	| Mbed OS version 	| Comments         	|
+|----------	|----------------	|--------------------------------------------------------	|-----------------	|------------------	|
+| 1.0      	| 20 September 2018	| David Saada ([@davidsaada](https://github.com/davidsaada/)) 	| 5.11+           	| Initial revision 	|
+
+# Introduction
+
+### Overview and background
+
+FileSystemStore is a lightweight implementation of the [KVStore](../KVStore/KVStore_design.md) interface over file systems.
+
+### Requirements and assumptions
+
+FileSystemStore assumes the underlying file system qualities for resilience and file validation. This means that if the underlying file system has no protection against power failures, then neither would FileSystemStore have.  
+When initializing this class, it is assumed that the underlying FileSystem is initialized and mounted. 
```

Comment:
OK, I think this is a key point to explain, which is also related to your comments below: One of the reasons for using this class is if the user wants to use POSIX APIs *in addition* to the get/set ones. This means that the file system is not dedicated to FileSystemStore, but rather hosts it. Using a dedicated file system for FileSystemStore would require a separate block device (or a slice), which adds overhead and wastes space assuming you want a "regular" file system as well.
So - this is why we need a dedicated directory to host the FileSystemStore files; this is why we assume the file system is already initialized; this is why we can't map the `reset` API to the file system `format` one and so on. If it wasn't clear from the design doc, I'll try to explain it more clearly.

Reply:
The SlicingBlockDevice is a good comparison. It provides organization, but comes with a code cost. So we provide it as an optional layer.

The current FileSystemStore also provides organization, which is a great feature, but it's not the primary goal of KVStore -> FileSystem. We need a way to mount a FileSystem as a KVStore without additional cost.

I think the best thing to do is introduce KVStore into the FileSystem inheritance chain (see the picture above), and keep the FileSystemStore as is (analogous to the SlicingBlockDevice).

---

> One of the reasons for using this class is if the user wants to use POSIX APIs in addition to the get/set ones. This means that the file system is not dedicated to FileSystemStore, but rather hosts it.

Sorry if I'm missing something, but I'm not sure how these statements relate. One example that has come up is MCC + firmware images (this was rejected for other reasons, but is a valid use case). In this case, a dedicated FileSystem/KVStore makes perfect sense. The KVStore API provides the storage for MCC, and the user can create the firmware images through the FileSystem APIs with a name that does not conflict with the KV pairs. Adding the layer for hosting complicated the user interface.

> Using a dedicated file system for FileSystemStore would require a separate block device (or a slice), which adds overhead and wastes space assuming you want a "regular" file system as well.

That's true. On the other hand, maybe a user wants their KV store isolated from the file system to avoid wear-amplification? Maybe they need their KV store to last a long time independently of the file system. Note it would still be advantageous to use a filesystem in both slices because of code sharing.

We don't need to out-think our users. We should be providing flexible options in the code and educating them on best practices in the docs.

Reply:
@geky I'm probably missing big time here. Maybe your drawing confuses me a bit, especially the line between `FileSystem` and `KVStore`. Do you suggest that `FileSystem` base class should be modified to start inheriting from `KVStore` now? 
In addition, I'm not sure what state we're holding in `FileSystemStore`. The containing directory name is a *fixed* one, with a unique name (say "$FileSystemStore$). It doesn't need any dynamic allocation whatsoever, nor hold a any state. I think I need more clarifications to this concept before I can go on and refer to the rest of your comments.

Reply:
Ah,

> Do you suggest that FileSystem base class should be modified to start inheriting from KVStore now?

Yep. This would tie together the two high-level storage classes. The layer from KVStore::get -> FileSystem::file_read can be implemented as default functions in the FileSystem base class. C++ compilers do a good job of optimizing these out if they are unused.

---

> The containing directory name is a _fixed_ one, with a unique name (say "$FileSystemStore$").

Ah thanks, I missed that. So it's impossible to have two independent FileSystemStore instances on a single file system?

> It doesn't need any dynamic allocation whatsoever, nor hold a any state.

How does FileSystemStore construct the new path for the key? Don't we need to convert from "key" to "$FileSystemStore$/key"?

---

The comparison of FileSystemStore and SlicingBlockDevice was with my incorrect assumption that FileSystemStore takes a user-provided directory name. In that case FileSystemStore provides an additional organization feature.

If not, I don't think FileSystemStore is necessary assuming this logic moves to the defaults in FileSystem.

Reply:
OK, I think I understand now.
I believe this approach has a few problems. You basically want to add all KVStore APIs to FileSystem class, having all file systems inheriting this functionality. In order for this to happen, you will need to implement all these APIs in the FileSysem class. However, FileSystem is just an interface class, meaning that ALL APIs will need implementation in this header file. These are no one liners. The iterator next API, for instance, requires a loop in order to see if the file name matches the prefix. Incremental set APIs have some logic to handle the handles (no pun intended) and so do the iterator ones. Offset argument in the get function also requires some logic, even if a simple one. This approach also makes me understand your other comments regarding metadata and dedicated directory in the wider context: they all get in the way of this striving for simplicity, as otherwise it would be unreasonable to have all this code in the interface class. While we can argue about the need for metadata and dedicated directory, this whole approach chains our legs. If for instance we will be required to add support for more flags (something our design supports now), we will either have to add code in the interface class, or look for the functionality in the underlying file systems. Write once is a good example. While I'm not sure it can be done in the existing ones, no one promises us it'll be supported in future file systems.
But that's not all of it. Adding this functionality to the base class, as suggested here, would have a much larger impact on all current file systems. We may add bugs there (no one is bug proof) and will certainly need to widen our current tests to find these potential bugs. I'm not even sure how to do it - we currently have separate tests for FatFS and LittleFS, so we'll have to restructure our tests to cover it, something I don't think we want to do (especially in this time frame).

Reply:
As for the dedicated directory: the way I see it, one needs to decide what we want - to dedicate the file system for key value storage or give the option not to. If you want it dedicated for KV storage, then what are the other POSIX APIs needed for? One can even use TDBStore instead. Number of keys is no obstacle (as you suggested below). You also mentioned above that the user may want to have separate storage blocks for different uses (file system vs. KVStore). That's true if the user has enough space for it. However, I can think of use cases they don't have this space (for instance if they only use internal flash), meaning that separating the file systems is not optimal. Current design offers the flexibility of either having separate file system for POSIX and KVStore usage or for using both in the same file system, which is the second option. Now, if we want our file system to support both, we certainly have to separate these physically, hence the KVStore directory. Otherwise, iterator actions would return files that are not of the KVStore type (like logs), while the reset action wouldn't know which files to delete.

Reply:
> Interface classes can have cpp files, FileSystem actually already has this along with several non-trivial defaults:

Sorry, missed this one.

> This is the same as adding FileSystemStore isn't it?

It's not the same. If we have FileSystem inherit KVStore, then all file systems will get the KVStore APIs by default, which is a change of behavior in current file systems, requiring tests and potentially adding bugs. Not the case with FileSystemStore.

Reply:
> As for the dedicated directory: the way I see it, one needs to decide what we want

Ah, and I don't think that should be us. I think our users should be the ones who decide.

My proposal is we provide both:
1. FileSystemStore - provides dedicated directory
2. FileSystem->KVStore inheritance - provides direct access to filesystem as a KVStore

To be clear, I don't think option 1 is required for 5.11 and could be pushed back to 5.12. I also don't think providing only option 1 is a good idea. I'm worried about the cost of the FileSystemStore layer, even if we drop the extra metadata, I don't see how we can avoid dynamic allocation to set the directory's prefix. Also all of the logic and tests for option 2 are required for option 1, so I don't think we gain much by not providing option 2.

> to dedicate the file system for key value storage

Note with option 2 you can still use POSIX + KV APIs. Going back to the firmware + KV storage, the user could store all KV pairs as "kv.something", and the firmware as "fw.firmware1". This would let them store different types of files on the same disk.

This also avoids the dynamic string costs as the names are simple constants stored in the upper layers.

The same problem happens if you have a SecureStore on TDBStore, now the user may have files with different levels of enc/auth/rbp.

> Otherwise, iterator actions would return files that are not of the KVStore type (like logs)

Isn't this the purpose of the pattern in the iterator functions?

> If you want it dedicated for KV storage, then what are the other POSIX APIs needed for?

There are reasons besides the API that users may want to use something other than TDBStore.

- Maybe user wants KV store on NAND and needs to avoid bad blocks on storage?
- Maybe user wants to use FRAM or SD cards and doesn't want the overhead of simulating flash?
- Maybe user wants to plug the SD card into his machine to debug?
- Maybe user has so many KV pairs the RAM table gets unwieldy? 
- Maybe user is already using a filesystem elsewhere in the system and doesn't want to pay double the code cost to store keys on a separate piece of storage?

I'm not saying won't use TDBStore, there are plenty of use cases where TDBStore is a better choice. But we need to give users that choice. IoT is a developing field, if we can't provide flexibility we'll just lose users to other OSs.

Reply:
Also consider if we end up adding enough features to TDBStore to where it can actually provide the POSIX APIs. I think the only way to introduce POSIX APIs in TDBStore in a backwards compatible would be by adding to the inheritance list.

Reply:
Oh sorry! GitHub doesn't update in realtime so I missed your response.

> It's not the same. If we have FileSystem inherit KVStore, then all file systems will get the KVStore APIs by default, which is a change of behavior in current file systems, requiring tests and potentially adding bugs. Not the case with FileSystemStore.

Adding KVStore APIs shouldn't change the FileSystem APIs, does it? The only testing change is the addition of the KVStore APIs for LittleFS and FAT, which is true if we add the FileSystemStore or FileSystem->KVStore inheritance.

Reply:
I think this is the sort of thing where code explains more than words. See https://github.com/davidsaada/mbed-os/pull/10.

We could have a really nicely bundled storage stack.